File: voice_ai_api/app/services/assembly_ai_service.rb
require 'net/http'
require 'json'

class AssemblyAIService
  BASE_URL = 'https://api.assemblyai.com/v2'
  
  def initialize(api_key:)
    @api_key = api_key
  end
  
  # Upload audio file and get transcription
  def transcribe_audio(audio_url:, language: 'en')
    # Step 1: Submit transcription request
    transcript = submit_transcription(audio_url: audio_url, language: language)
    transcript_id = transcript['id']
    
    # Step 2: Poll for completion
    wait_for_transcription(transcript_id: transcript_id)
  end
  
  # Submit transcription job
  def submit_transcription(audio_url:, language: 'en')
    uri = URI("#{BASE_URL}/transcript")
    
    request_body = {
      audio_url: audio_url,
      language_code: language
    }
    
    headers = {
      'authorization' => @api_key,
      'content-type' => 'application/json'
    }
    
    response = make_request(uri, :post, request_body, headers)
    JSON.parse(response.body)
  end
  
  # Get transcription status
  def get_transcription_status(transcript_id:)
    uri = URI("#{BASE_URL}/transcript/#{transcript_id}")
    
    headers = {
      'authorization' => @api_key
    }
    
    response = make_request(uri, :get, nil, headers)
    JSON.parse(response.body)
  end
  
  # Wait for transcription completion (polling)
  def wait_for_transcription(transcript_id:, max_wait: 300, poll_interval: 1)
    start_time = Time.now
    
    loop do
      transcript = get_transcription_status(transcript_id: transcript_id)
      
      if transcript['status'] == 'completed'
        return transcript['text'] || ''
      elsif transcript['status'] == 'error'
        raise "Transcription failed: #{transcript['error']}"
      elsif Time.now - start_time > max_wait
        raise "Transcription timeout"
      end
      
      sleep poll_interval
    end
  end
  
  # Upload local file to AssemblyAI (if needed)
  def upload_file(file_path:)
    uri = URI("#{BASE_URL}/upload")
    
    headers = {
      'authorization' => @api_key
    }
    
    # Read file and upload
    File.open(file_path, 'rb') do |file|
      request = Net::HTTP::Post.new(uri.path)
      headers.each { |key, value| request[key] = value }
      request.body = file.read
      
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true
      response = http.request(request)
      
      unless response.is_a?(Net::HTTPSuccess)
        raise "AssemblyAI Upload Error: #{response.code} - #{response.body}"
      end
      
      result = JSON.parse(response.body)
      return result['upload_url']
    end
  end
  
  private
  
  def make_request(uri, method, body, headers)
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = true
    
    request = case method
              when :post
                Net::HTTP::Post.new(uri.path)
              when :get
                Net::HTTP::Get.new(uri.path)
              end
    
    headers.each { |key, value| request[key] = value }
    request.body = body.to_json if body
    
    response = http.request(request)
    
    unless response.is_a?(Net::HTTPSuccess)
      raise "AssemblyAI API Error: #{response.code} - #{response.body}"
    end
    
    response
  end
end

File: VoiceAIApp/services/AssemblyAIService.ts

export interface AssemblyAIConfig {
  apiKey: string;
}

/**
 * AssemblyAI STT servisi
 * Ücretsiz tier ile aylık 5 saat transkripsiyon
 */
export class AssemblyAIService {
  private apiKey: string;
  private baseUrl = 'https://api.assemblyai.com/v2';

  constructor(config: AssemblyAIConfig) {
    this.apiKey = config.apiKey;
  }

  async transcribeAudio(audioUri: string): Promise<string> {
    try {
      console.log('Starting AssemblyAI transcription...');
      console.log('Audio URI:', audioUri);

      // React Native FormData supports URI directly
      const formData = new FormData();
      formData.append('audio', {
        uri: audioUri,
        type: 'audio/m4a',
        name: 'recording.m4a',
      } as any);

      console.log('Uploading audio to AssemblyAI...');

      // Upload audio first
      const uploadResponse = await fetch(`${this.baseUrl}/upload`, {
        method: 'POST',
        headers: {
          'authorization': this.apiKey
        },
        body: formData,
      });

      if (!uploadResponse.ok) {
        const errorText = await uploadResponse.text();
        console.error('AssemblyAI upload error:', errorText);
        throw new Error(`Upload failed: ${uploadResponse.status} - ${errorText}`);
      }

      const uploadResult = await uploadResponse.json();
      const uploadUrl = uploadResult.upload_url;
      console.log('Audio uploaded, URL:', uploadUrl);

      // Submit for transcription
      const transcribeResponse = await fetch(`${this.baseUrl}/transcript`, {
        method: 'POST',
        headers: {
          'authorization': this.apiKey,
          'content-type': 'application/json'
        },
        body: JSON.stringify({
          audio_url: uploadUrl,
          language_code: 'en'
        }),
      });

      if (!transcribeResponse.ok) {
        const errorText = await transcribeResponse.text();
        console.error('AssemblyAI transcription error:', errorText);

        if (transcribeResponse.status === 401 || transcribeResponse.status === 403) {
          throw new Error('Invalid AssemblyAI API key or insufficient credits. Please check your API key.');
        }

        throw new Error(`Transcription failed: ${transcribeResponse.status} - ${errorText}`);
      }

      const transcribeResult = await transcribeResponse.json();
      const transcriptId = transcribeResult.id;
      console.log('Transcription submitted, ID:', transcriptId);

      // Poll for completion
      let transcript = transcribeResult;
      while (transcript.status === 'queued' || transcript.status === 'processing') {
        await new Promise(resolve => setTimeout(resolve, 1000));
        const statusResponse = await fetch(`${this.baseUrl}/transcript/${transcriptId}`, {
          headers: {
            'authorization': this.apiKey
          }
        });
        transcript = await statusResponse.json();
        console.log('Transcription status:', transcript.status);
      }

      if (transcript.status === 'error') {
        throw new Error('Transcription failed: ' + transcript.error);
      }

      // Extract text
      const text = transcript.text || '';

      if (!text || text.trim() === '') {
        throw new Error('Empty transcript received from AssemblyAI');
      }

      console.log('Transcription complete:', text);
      return text;
    } catch (error) {
      console.error('Error in AssemblyAI transcription:', error);
      throw error;
    }
  }
}

